text/trace/0157-messaging-semantic-conventions.md

```diff
@@ -0,0 +1,184 @@
+| An OTEP proposing a set of attributes and conventions covering the scenarios in this document is approved and merged. | 10/29/2021 |
+| Proposed specification changes are verified by prototypes for the scenarios and examples below. | 11/15/2021 |
+| The [specification for messaging semantic conventions for tracing](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md) are updated according to the OTEP mentioned above and are declared [stable](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md#stable). | 11/30/2021 |
+
+## Terminology
+
+### Message
+
+A "message" is a transport wrapper for the transfer of information between two
+or more parties. The information is a combination of data and metadata.
+Messages may be transferred directly between parties or via one or more
+intermediaries.  A message may carry annotations that are directed at
+intermediaries on the message path. Messages are uniquely identifiable.
+
+In the strict sense, a _message_ is an item of data that is sent to a specific
+destination, whereas an _event_ is a signal emitted by a component upon
+reaching a given state. This document is agnostic of those differences and uses
+the term "message" in a wider sense to cover both concepts.
+
+### Producer
+
+The "producer" is a specific instance, process or device that creates and
+publishes a message. "Publishing" is the process of sending a message or batch
+to the intermediary or consumer.
+
+### Consumer
+
+A "consumer" receives the message and acts upon it. It uses the context and
+data to execute some logic, which might lead to the occurrence of new events.
+
+The consumer receives, processes, and settles a message. "Receiving" is the
+process of obtaining a message from the intermediary, "processing" is the
+process of acting on the information a message contains, "settling" is the
+process of notifying an intermediary that a message was processed successfully.
+
+### Intermediary
+
+An "intermediary" receives a message to forward it to the next receiver, which
+might be another intermediary or a consumer.
+
+## Scenarios
+
+Producing and consuming a message involves five stages:
+
+```
+PRODUCER
+
+Create
+  |                            CONSUMER
+  v        +--------------+                   
+Publish -> | INTERMEDIARY | -> Receive
+           +--------------+       |
+                  ^               v
+                  .            Process
+                  .               |
+                  .               v
+                  . . . . . .  Settle
+```
+
+1. The producer creates a message.
+2. The producer publishes the message to an intermediary.
```

Comment:
What is not present in this OTEP but defined in the current semantic conventions is the concept of the "(logical) message destination" of which the current conventions know the two kinds "queue" (1 to 1) and "topic" (1 to many).
Or maybe the concept of "intermediary" is supposed to capture not only the concept of a "broker" (aka "physical message destination" in the current conventions) but also the logical message destination?
Would it make sense to split the Intermediary up?

Quoting the current conventions:

> A message is sent by a message *producer* to:
>
> * Physically: some message *broker* (which can be e.g., a single server, or a cluster, or a local process reached via IPC). The broker handles the actual delivery, re-delivery, persistence, etc. In some messaging systems the broker may be identical or co-located with (some) message consumers.
With Apache Kafka, the physical broker a message is written to depends on the number of partitions, and which broker is the *leader* of the partition the record is written to.
> * Logically: some particular message *destination*.

Reply:
> What is not present in this OTEP but defined in the current semantic conventions is the concept of the "(logical) message destination" of which the current conventions know the two kinds "queue" (1 to 1) and "topic" (1 to many).

I wonder how much this overlaps with the distinct scenarios of individual settlement and checkpoint settlement described below. Let's discuss this in the next meeting.

However,  I think whether one publishes to a queue or to a topic doesn't alter the model of the 5 stages. This can rather be captured in different scenarios that are based on this model.

Reply:
I agree that it makes no fundamental difference with what steps are involved, although there are probably many differences if you look a bit closer.
I was less concerned about the steps and more about the entities involved. E.g. if there is a problem with message delivery, it might be because some configured max queue (logical destination) length is exceeded. Or there might be a problem with the "physical" server you send the message to (network overloaded, etc). I think the model should capture not only the  steps performed, but also the important entities that are involved in each step.
That said, I think it _could_ make sense to focus on the steps first, because these involve span relationships / context propagation / ... which are probably more difficult to get right than the entitities which will probably come down to some span attributes here and there.

Reply:
> That said, I think it could make sense to focus on the steps first, because these involve span relationships / context propagation / ... which are probably more difficult to get right than the entitities which will probably come down to some span attributes here and there.

@Oberon00 When you talk about entities, do you have something in mind that goes beyond the conceptual producer/intermediary/consumer entities? I'm wondering whether we can define entities more fine-grained in such a high-model, as more fine-grained distinctions might be specific to different messaging systems.